driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java

```diff
@@ -29,357 +29,301 @@
+     * @throws NullPointerException if {@code bytes} is {@code null}
+     */
+    public abstract T deserialize(ByteBuffer bytes, ProtocolVersion protocolVersion) throws InvalidTypeException;
+
+    /**
+     * Parse the given CQL string into an instance of the Java type
+     * handled by this codec.
+     * @param value CQL string
+     * @return An instance of T
+     * @throws InvalidTypeException if the given value cannot be parsed into the expected type
+     * @throws NullPointerException if {@code value} is {@code null}
+     */
+    public abstract T parse(String value) throws InvalidTypeException;
+
+    /**
+     * Format the given value as a valid CQL string according
+     * to the CQL type handled by this codec.
+     * @param value An instance of T
+     * @return CQL string
+     * @throws InvalidTypeException if the given value does not have the expected type
+     * @throws NullPointerException if {@code value} is {@code null}
+     */
+    public abstract String format(T value) throws InvalidTypeException;
 
-    @SuppressWarnings("unchecked")
-    static <T> TypeCodec<T> createFor(DataType.Name name) {
-        assert !name.isCollection();
-        return (TypeCodec<T>)primitiveCodecs.get(name);
+    /**
+     * Return {@code true} if this codec is capable of serializing
+     * the given {@code javaType}.
+     *
+     * @param javaType The Java type this codec should serialize from and deserialize to; cannot be {@code null}.
+     * @return {@code true} if the codec is capable of serializing
+     * the given {@code javaType}, and {@code false} otherwise.
+     * @throws NullPointerException if {@code javaType} is {@code null}.
+     */
+    public boolean accepts(TypeToken javaType) {
+        checkNotNull(javaType);
+        if(javaType.isPrimitive()) {
+            javaType = primitiveToWrapperMap.get(javaType);
+        }
+        return this.javaType.isAssignableFrom(javaType);
     }
 
-    @SuppressWarnings("unchecked")
-    static <T> TypeCodec<List<T>> listOf(DataType arg, ProtocolVersion protocolVersion) {
-        PrimitiveCollectionCodecs codecs = PrimitiveCollectionCodecs.forVersion(protocolVersion);
-        TypeCodec<List<?>> codec = codecs.primitiveListsCodecs.get(arg.getName());
-        return codec != null ? (TypeCodec)codec : new ListCodec<Object>(arg.codec(protocolVersion), protocolVersion);
+    /**
+     * Return {@code true} if this codec is capable of deserializing
+     * the given {@code cqlType}.
+     *
+     * @param cqlType The CQL type this codec should deserialize from and serialize to; cannot be {@code null}.
+     * @return {@code true} if the codec is capable of deserializing
+     * the given {@code cqlType}, and {@code false} otherwise.
+     * @throws NullPointerException if {@code javaType} is {@code null}.
+     */
+    public boolean accepts(DataType cqlType) {
+        checkNotNull(cqlType);
+        // text is merely an alias for varchar
+        if (cqlType == text())
```

Comment:
Ran into a problem where I had a codec of frozen<tuple<timestamp,text> <-> org.joda.time.DateTime and it refused to map because the cqlType from cassandra was frozen<tuple<timestamp,varchar> (even though it is declared as text in my cql).  I think we should document this possibility (as it doesn't sound very easy to handle deterministically unless i'm mistaken).  It's easy to work around (always use DataType.varchar in your codec instead of text)